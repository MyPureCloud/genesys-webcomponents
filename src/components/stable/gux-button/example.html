<h4>Enabled</h4>
<div>
  <gux-button title="default">Default</gux-button>
  <gux-button title="Primary" accent="primary">Primary</gux-button>
  <gux-button title="Secondary" accent="secondary">Secondary</gux-button>
  <gux-button title="Invalid accent" accent="invalid"
    >Invalid accent</gux-button
  >
  <gux-button title="Slotted element" accent="invalid">
    <span>Text in a span element</span>
  </gux-button>
  <gux-button title="Slotted element" accent="invalid">
    <span>Text in a span element</span>
    <span>More text in a span element</span>
  </gux-button>
</div>

<h4>Disabled</h4>
<div>
  <gux-button title="default" disabled>Default</gux-button>
  <gux-button title="Primary" accent="primary" disabled>Primary</gux-button>
  <gux-button title="Secondary" accent="secondary" disabled
    >Secondary</gux-button
  >
  <gux-button title="Invalid accent" accent="invalid" disabled
    >Invalid accent</gux-button
  >
  <gux-button title="Slotted element" accent="invalid" disabled>
    <span>Text in a span element</span>
  </gux-button>
  <gux-button title="Slotted element" accent="invalid" disabled>
    <span>Text in a span element</span>
    <span>More text in a span element</span>
  </gux-button>
</div>

<style>
  gux-button {
    margin: 10px;
  }
</style>
