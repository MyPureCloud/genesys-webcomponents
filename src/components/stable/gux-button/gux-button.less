@import (reference) '../../../style/color.less';
@import (reference) '../../../style/mixins.less';

// Variables part
@primary-dark-background-color: @gux-blue-60;
@primary-dark-border-color: @gux-black-30;
@primary-dark-color: @gux-type-dark;
@primary-dark-hover-background-color: @expalette-primary-dark-hover-background-color;
@secondary-dark-background-color: @expalette-background-color-dark;
@secondary-dark-border-color: @gux-black-30;
@secondary-dark-color: @gux-type-dark;
@secondary-dark-hover-background-color: @expalette-secondary-dark-hover-background-color;

gux-button {
  display: inline-block;
  min-width: 32px;
  pointer-events: none;

  > button {
    width: 100%;
    padding: 0 16px;
    height: 32px;
    border-radius: 2px;
    border: none;
    cursor: pointer;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    font-size: 12px;
    font-weight: bold;
    pointer-events: auto;

    &[disabled] {
      opacity: 0.5;
      cursor: default;
    }

    &:focus {
      outline: none;
    }
  }
}

// Theming
.gux-button-dark-theme {
  > button {
    background-color: @secondary-dark-background-color;
    border: 1px solid @secondary-dark-border-color;
    color: @secondary-dark-color;

    &:hover {
      background-color: @secondary-dark-hover-background-color;
    }

    &:active {
      box-shadow: 0 0 0 3px @gux-blue-80;
    }

    &:focus {
      box-shadow: 0 0 0 3px fade(@gux-blue-80, 50%);
    }

    &[disabled] {
      &:hover {
        background-color: @secondary-dark-background-color;
      }

      &:active {
        box-shadow: none;
      }

      &:focus {
        box-shadow: none;
      }
    }

    &.gux-primary {
      background-color: @primary-dark-background-color;
      border: 1px solid @primary-dark-border-color;
      color: @primary-dark-color;

      &:hover {
        background-color: @primary-dark-hover-background-color;
      }

      &:active {
        box-shadow: 0 0 0 3px @gux-blue-80;
      }

      &:focus {
        box-shadow: 0 0 0 3px fade(@gux-blue-80, 50%);
      }

      &[disabled] {
        &:hover {
          background-color: @primary-dark-background-color;
        }

        &:active {
          box-shadow: none;
        }

        &:focus {
          box-shadow: none;
        }
      }
    }
  }
}

.gux-button-light-theme {
  > button {
    background-color: @gux-grey-60;
    color: @gux-black-50;

    &:hover:enabled {
      background-color: @gux-grey-50;
    }

    &:active:enabled {
      background-color: @gux-grey-40;
    }

    &:focus:enabled {
      .focus-ring();
    }

    &.gux-tertiary {
      background-color: @gux-grey-100;
      color: @gux-blue-60;
      border: 1px solid @gux-blue-60;

      &:hover:enabled {
        background-color: @gux-blue-50;
        color: @gux-grey-100;
      }

      &:active:enabled {
        background-color: @gux-blue-40;
        color: @gux-grey-100;
      }
    }

    &.gux-primary {
      background-color: @gux-blue-60;
      color: @gux-grey-100;

      &:hover:enabled {
        background-color: @gux-blue-50;
      }

      &:active:enabled {
        background-color: @gux-blue-40;
      }
    }
  }
}

// Default Theme
gux-button {
  .gux-button-light-theme();
}

// Dark
.gux-dark-theme {
  gux-button {
    .gux-button-dark-theme();
  }
}

gux-button.gux-dark-theme {
  .gux-button-dark-theme();
}

// Light
.gux-light-theme {
  gux-button {
    .gux-button-light-theme();
  }
}

gux-button.gux-light-theme {
  .gux-button-light-theme();
}
