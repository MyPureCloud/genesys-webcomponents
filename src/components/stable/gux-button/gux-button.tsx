import {
  Component,
  ComponentInterface,
  Element,
  h,
  JSX,
  Prop
} from '@stencil/core';

import { logComponentUsage } from '../../../common/log-component-usage';

export type GuxButtonAccent = 'primary' | 'secondary';

@Component({
  styleUrl: 'gux-button.less',
  tag: 'gux-button'
})
export class GuxButton implements ComponentInterface {
  @Element()
  private root: HTMLElement;

  /**
   * The component title
   */
  @Prop()
  title: string;

  /**
   * Indicate if the button is disabled or not
   */
  @Prop()
  disabled = false;

  /**
   * The component accent (secondary or primary).
   */
  @Prop()
  accent: GuxButtonAccent = 'secondary';

  constructor() {
    logComponentUsage('gux-button');
  }

  componentWillLoad() {
    this.makeSlotContentDisableable();
  }

  render(): JSX.Element {
    return (
      <button
        title={this.title}
        disabled={this.disabled}
        class={this.accent === 'primary' ? 'gux-primary' : 'gux-secondary'}
      >
        <slot />
      </button>
    );
  }

  private makeSlotContentDisableable() {
    Array.from(this.root.children).forEach(slotElement => {
      slotElement.addEventListener('click', (event: MouseEvent): void => {
        if (this.disabled) {
          event.stopImmediatePropagation();
          event.stopPropagation();
          event.preventDefault();
        }
      });
    });
  }
}
