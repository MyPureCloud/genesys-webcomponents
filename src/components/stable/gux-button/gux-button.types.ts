export type GuxButtonAccent = 'primary' | 'secondary' | 'tertiary';

export type GuxButtonType = 'button' | 'submit' | 'reset';
