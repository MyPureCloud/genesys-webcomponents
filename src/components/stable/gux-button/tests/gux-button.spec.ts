import { newSpecPage } from '@stencil/core/testing';
import { GuxButton } from '../../gux-button/gux-button';

const components = [GuxButton];
const language = 'en';

describe('gux-modal', () => {
  describe('#render', () => {
    [
      {
        description: 'should render default button',
        html: '<gux-button title="default">Button</gux-button>'
      },
      {
        description: 'should render primary button',
        html: '<gux-button title="Primary" accent="primary">Button</gux-button>'
      },
      {
        description: 'should render secondary button',
        html:
          '<gux-button title="Secondary" accent="secondary">Button</gux-button>'
      },
      {
        description: 'should render tertiary button',
        html:
          '<gux-button title="Tertiary" accent="tertiary">Button</gux-button>'
      },
      {
        description: 'should render invalid accent button',
        html:
          '<gux-button title="Invalid accent" accent="invalid">Invalid accent</gux-button>'
      },
      {
        description: 'should render disabled default button',
        html: '<gux-button title="default" disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled primary button',
        html:
          '<gux-button title="Primary" accent="primary" disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled secondary button',
        html:
          '<gux-button title="Secondary" accent="secondary" disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled tertiary button',
        html:
          '<gux-button title="Tertiary" accent="tertiary" disabled>Button</gux-button>'
      },
      {
        description: 'should render disabled invalid accent button',
        html:
          '<gux-button title="Invalid accent" accent="invalid" disabled>Invalid accent</gux-button>'
      }
    ].forEach(({ description, html }) => {
      it(description, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.rootInstance).toBeInstanceOf(GuxButton);
        expect(page.root).toMatchSnapshot();
      });
    });
  });

  describe('click', () => {
    it('should fire a click event when an enabled button is clicked', async () => {
      const html = '<gux-button title="default">Button</gux-button>';
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const button = page.root.querySelector('button') as HTMLButtonElement;
      const clickSpy = jest.fn();

      element.addEventListener('click', clickSpy);

      button.click();
      await page.waitForChanges();

      expect(clickSpy).toHaveBeenCalled();
    });

    it('should fire a click event when an enabled button slot content is clicked', async () => {
      const html = '<gux-button title="default"><span>Span</span></gux-button>';
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const span = page.root.querySelector('span') as HTMLButtonElement;
      const clickSpy = jest.fn();

      element.addEventListener('click', clickSpy);

      span.click();
      await page.waitForChanges();

      expect(clickSpy).toHaveBeenCalled();
    });

    it('should not fire a click event when a disabled button slot content is clicked', async () => {
      const html =
        '<gux-button title="default" disabled><span>Span</span></gux-button>';
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const span = page.root.querySelector('span') as HTMLButtonElement;
      const clickSpy = jest.fn();

      element.addEventListener('click', clickSpy);

      span.click();
      await page.waitForChanges();

      expect(clickSpy).not.toHaveBeenCalled();
    });
  });
});
