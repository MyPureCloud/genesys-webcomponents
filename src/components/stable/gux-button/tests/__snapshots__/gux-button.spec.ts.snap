// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-modal #render should render default button 1`] = `
<gux-button title="default">
  <!---->
  <button class="gux-secondary" title="default" type="button">
    Button
  </button>
</gux-button>
`;

exports[`gux-modal #render should render disabled default button 1`] = `
<gux-button disabled="" title="default">
  <!---->
  <button class="gux-secondary" disabled="" title="default" type="button">
    Button
  </button>
</gux-button>
`;

exports[`gux-modal #render should render disabled invalid accent button 1`] = `
<gux-button accent="invalid" disabled="" title="Invalid accent">
  <!---->
  <button class="gux-invalid" disabled="" title="Invalid accent" type="button">
    Invalid accent
  </button>
</gux-button>
`;

exports[`gux-modal #render should render disabled primary button 1`] = `
<gux-button accent="primary" disabled="" title="Primary">
  <!---->
  <button class="gux-primary" disabled="" title="Primary" type="button">
    Button
  </button>
</gux-button>
`;

exports[`gux-modal #render should render disabled secondary button 1`] = `
<gux-button accent="secondary" disabled="" title="Secondary">
  <!---->
  <button class="gux-secondary" disabled="" title="Secondary" type="button">
    Button
  </button>
</gux-button>
`;

exports[`gux-modal #render should render disabled tertiary button 1`] = `
<gux-button accent="tertiary" disabled="" title="Tertiary">
  <!---->
  <button class="gux-tertiary" disabled="" title="Tertiary" type="button">
    Button
  </button>
</gux-button>
`;

exports[`gux-modal #render should render invalid accent button 1`] = `
<gux-button accent="invalid" title="Invalid accent">
  <!---->
  <button class="gux-invalid" title="Invalid accent" type="button">
    Invalid accent
  </button>
</gux-button>
`;

exports[`gux-modal #render should render primary button 1`] = `
<gux-button accent="primary" title="Primary">
  <!---->
  <button class="gux-primary" title="Primary" type="button">
    Button
  </button>
</gux-button>
`;

exports[`gux-modal #render should render secondary button 1`] = `
<gux-button accent="secondary" title="Secondary">
  <!---->
  <button class="gux-secondary" title="Secondary" type="button">
    Button
  </button>
</gux-button>
`;

exports[`gux-modal #render should render tertiary button 1`] = `
<gux-button accent="tertiary" title="Tertiary">
  <!---->
  <button class="gux-tertiary" title="Tertiary" type="button">
    Button
  </button>
</gux-button>
`;
